src/Select.purs

```diff
@@ -90,266 +77,300 @@ derive instance ordVisibility :: Ord Visibility
+-- | distinguishes one event type from another.
+data SelectEvent
+  = NewSearch String
+  | VisibilityChangedTo Boolean
+  | SelectedIndex Int
+
+newtype UseSelect hooks = UseSelect
+  (UseDebouncer String
+  (UseState SelectState hooks))
+
+derive instance newtypeUseSelect :: Newtype (UseSelect hooks) _
 
-type Spec' st input m = Spec st (Const Void) Void () input Void m
-
-defaultSpec
-  :: forall st query action slots input msg m
-   . Spec st query action slots input msg m
-defaultSpec =
-  { render: const (HH.text mempty)
-  , handleAction: const (pure unit)
-  , handleQuery: const (pure Nothing)
-  , handleEvent: const (pure unit)
-  , receive: const Nothing
-  , initialize: Nothing
-  , finalize: Nothing
+-- | A `SelectInput` value whose defaults can be overrided. **Note**:
+-- | `getItemCount` must be overrided:
+-- |
+-- | Default values are:
+-- | ```
+-- | { inputType: Toggle
+-- | , search: Nothing
+-- | , debounceTime: Nothing
+-- | , getItemCount: pure 0 -- this must be overrided!
+-- | , pushNewSearch: \_ -> pure unit
+-- | , pushVisibilityChanged: \_ -> pure unit
+-- | , pushSelectedIdxChanged: \_ -> pure unit
+-- | }
+-- | ```
+-- |
+-- | Example:
+-- | ```
+-- | events <- useEvent
+-- | SelectReturn select <- useSelect $ selectInput
+-- |   { getItemCount = pure (length items)
+-- |   , pushNewSearch = events.push
+-- |   }
+-- | ```
+selectInput :: forall m. SelectInput m
+selectInput =
+  { inputType: Toggle
+  , search: Nothing
+  , debounceTime: Nothing
+  , getItemCount: pure 0
+  , pushNewSearch: mempty
+  , pushVisibilityChanged: mempty
+  , pushSelectedIdxChanged: mempty
   }
 
-component
-  :: forall st query action slots input msg m
+useSelect
+  :: forall m
    . MonadAff m
-  => Row.Lacks "debounceRef" st
-  => Row.Lacks "visibility" st
-  => Row.Lacks "highlightedIndex" st
-  => (input -> Input st)
-  -> Spec st query action slots input msg m
-  -> H.Component HH.HTML (Query query slots) input msg m
-component mkInput spec = H.mkComponent
-  { initialState: initialState <<< mkInput
-  , render: spec.render
-  , eval: H.mkEval
-      { handleQuery: handleQuery spec.handleQuery
-      , handleAction: handleAction spec.handleAction spec.handleEvent
-      , initialize: Just (Initialize spec.initialize)
-      , receive: map Action <<< spec.receive
-      , finalize: map Action spec.finalize
+  => SelectInput m
+  -> Hook m UseSelect (SelectReturn m)
+useSelect inputRec =
+  let
+    initialSearchValue = fromMaybe "" inputRec.search
+    debounceTime = fromMaybe mempty inputRec.debounceTime
+  in Hooks.wrap Hooks.do
+    state /\ stateId <- useState
+      { search: initialSearchValue
+      , visibility: Off
+      , highlightedIndex: Nothing
+      }
+
+    searchDebouncer <- useDebouncer debounceTime \lastSearchState -> Hooks.do
+      case inputRec.inputType of
+        Text -> do
+          -- modifyState (_ { highlightedIndex = (Just 0) })
```

Comment:
Should this be uncommented?

Reply:
Nope! I swtiched `modifyState` to `H.modify_` and it took a while to realize I needed `Hooks.modify_` not the other one. I likely commented this out, thinking there was another issue elsewhere and then forgot to uncomment it when I found out what the real problem was.

Reply:
Fixed in latest commits.